// pro-gallery/dist/esm/components/gallery/index.js
import { __assign, __extends, __rest } from "tslib";
import React from 'react';
import { defaultStyles, BlueprintsManager, GALLERY_CONSTS, utils, } from 'pro-gallery-lib';
import ProGallery from './proGallery/proGallery';
import basePropTypes from './proGallery/propTypes';
var BaseGallery = /** @class */ (function (_super) {
    __extends(BaseGallery, _super);
    function BaseGallery(props) {
        var _this = _super.call(this) || this;
        _this.blueprintsManager = new BlueprintsManager({ id: 'layoutingGallery' });
        _this.domId = props.domId || 'default-dom-id';
        _this.state = {
            blueprint: _this.blueprintsManager.existingBlueprint || null,
        };
        _this.blueprintsManager.init({
            api: {
                isUsingCustomInfoElements: function () {
                    return props.customHoverRenderer ||
                        props.customInfoRenderer ||
                        props.customSlideshowInfoRenderer;
                },
                fetchMoreItems: function (from) {
                    typeof props.eventsListener === 'function' &&
                        props.eventsListener(GALLERY_CONSTS.events.NEED_MORE_ITEMS, from);
                },
                onBlueprintReady: function (_a) {
                    var blueprint = _a.blueprint, blueprintChanged = _a.blueprintChanged;
                    if (blueprintChanged) {
                        _this.setBlueprint(blueprint);
                    }
                    else {
                        if (utils.isVerbose()) {
                            console.count('>>> Blueprint not changed, not setting it');
                        }
                    }
                },
            },
        });
        _this.onNewProps(props);
        return _this;
    }
    BaseGallery.prototype.setBlueprint = function (blueprint) {
        this.setState({ blueprint: blueprint });
    };
    BaseGallery.prototype.onNewProps = function (props) {
        var _this = this;
        var styles = props.styles, options = props.options, styleParams = props.styleParams, eventsListener = props.eventsListener, otherProps = __rest(props, ["styles", "options", "styleParams", "eventsListener"]);
        var _eventsListener = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var eventName = args[0], value = args[1];
            if (eventName === GALLERY_CONSTS.events.NEED_MORE_ITEMS) {
                _this.blueprintsManager.getMoreItems(value);
            }
            else {
                typeof eventsListener === 'function' && eventsListener.apply(void 0, args);
            }
        };
        var _styles = __assign(__assign(__assign(__assign({}, defaultStyles), options), styles), styleParams);
        this.galleryProps = __assign(__assign({}, otherProps), { styles: _styles, eventsListener: _eventsListener, domId: this.domId });
        this.blueprintsManager.createBlueprint(this.galleryProps).catch(function (e) {
            //TODOVER3 check totalItemsCount
            console.error('Could not breate a blueprints in layoutingIndex from given props', e);
        });
    };
    BaseGallery.prototype.UNSAFE_componentWillReceiveProps = function (newProps) {
        this.onNewProps(newProps);
    };
    BaseGallery.prototype.render = function () {
        var blueprint = this.state.blueprint;
        if (blueprint && Object.keys(blueprint).length > 0) {
            return React.createElement(ProGallery, __assign({}, this.galleryProps, blueprint));
        }
        else {
            return null;
        }
    };
    BaseGallery.propTypes = basePropTypes;
    return BaseGallery;
}(React.Component));
export default BaseGallery;
